### PR TITLE
[XLA:GPU] Adding an option that allocates temp buffer through separate memory allocator. 

### PR DESCRIPTION
**Pull request by shawnwang18**
This PR will assign XLA temp buffer to a dedicated memory space, and uses a standalone cuda async allocator to allocate temp buffer, as temp buffer is always allocated & deallocated for each step, this will ensure that temp buffer can be allocated to a stable address, which is good for cuda-graph's perf.

### CODE REVIEW

File: xla/debug_options_flags.cc [hunk 1]
```diff
@@ -1232,6 +1234,13 @@ void MakeDebugOptionsFlags(std::vector<tsl::Flag>* flag_list,
       "Enables NCCL User Buffer Registration. collective_memory_size in the "
       "allocator config must also be set to a non-zero value that is large "
       "enough to meet peak collective memory usage."));
+  flag_list->push_back(tsl::Flag(
+      "xla_gpu_enable_temp_async_buffer",
+      bool_setter_for(&DebugOptions::set_xla_gpu_enable_temp_async_buffer),
+      debug_options->xla_gpu_enable_temp_async_buffer(),
+      "Enables temp User Buffer Registration. Enable this flag will use a "
+      "separate cuda async memory allocator to allocate temp buffer, this will "
+      "allocate temp buffer to the fix address on every iteration"));
```
Comment (cheshire) [inline xla/debug_options_flags.cc:1243]: Nit: "fixed"

File: xla/pjrt/gpu/se_gpu_pjrt_client.cc [hunk 2]
```diff
@@ -789,59 +790,55 @@ namespace {
...
 
-absl::StatusOr<std::vector<se::MultiDeviceAdapter::AllocatorInfo>>
-CreateCudaAsyncAllocator(
-    se::Platform* platform,
-    const std::map<int, std::unique_ptr<LocalDeviceState>>& addressable_devices,
-    double memory_fraction, bool preallocate) {
-  CHECK_GT(addressable_devices.size(), 0);
-  std::vector<se::MultiDeviceAdapter::AllocatorInfo> allocators;
+absl::StatusOr<std::unique_ptr<se::GpuCudaMallocAsyncAllocator>>
+CreateCudaAsyncAllocator(const LocalDeviceState& device, double memory_fraction,
```
Comment (cheshire) [inline xla/pjrt/gpu/se_gpu_pjrt_client.cc:794]: Nit: could we separate changes to async allocator into a separate PR?

File: xla/service/buffer_assignment.cc [hunk 3]
```diff
@@ -251,9 +251,9 @@ void BufferAllocation::AddAssignment(const HloValue& buffer, int64_t offset,
   CHECK_LE(offset + size, size_)
       << "LogicalBuffer " << buffer
       << " size out of range at offset: " << offset << " with size: " << size;
-  CHECK_EQ(buffer.color(), color())
-      << "Buffer color " << buffer.color() << " for buffer " << buffer
-      << " does not match allocation color " << color() << ".";
+  // CHECK_EQ(buffer.color(), color())
```
Comment (cheshire) [inline xla/service/buffer_assignment.cc:254]: Should this be commented out?

File: xla/service/buffer_assignment.h [hunk 4]
```diff
@@ -570,7 +573,8 @@ class BufferAssignment {
 
   // Combines allocations of temporary buffers into one big BufferAllocation.
   void CombineTempAllocations(
-      const absl::flat_hash_set<BufferValue::Color>& private_stack_colors);
+      const absl::flat_hash_set<BufferValue::Color>& private_stack_colors,
+      bool enable_temp_user_buffer);
```
Comment (cheshire) [inline xla/service/buffer_assignment.h:577]: Is this a GPU-specific option? Should it go into a GPU subclass?

**Review by cheshire — CHANGES_REQUESTED** [review]
Could you add a more high-level description to the PR + tests?

File: xla/xla.proto [hunk 5]
```diff
@@ -786,7 +786,12 @@ message DebugOptions {
   // a deterministic implementation.
   bool xla_gpu_exclude_nondeterministic_ops = 297;
 
-  // Next id: 300
+  // Enables temp User Buffer Registration. Enable this flag will use a 
+  // separate cuda async memory allocator to allocate temp buffer, this will 
+  // allocate temp buffer to the fix address on every iteration.
+  bool xla_gpu_enable_temp_async_buffer = 301;
+
+  // Next id: 302
```
Comment (tdanyluk) [inline xla/xla.proto:794]: I think that I roughly understand this but the wording is not really clear to me.
IIUC the goal is to allocate a temp buffer which stays the same on every iteration? Why is it important that the buffer is "async"? 

Shouldn't we call this flag something that emphasizes the stable nature of the buffer?

File: xla/service/buffer_assignment.h [hunk 6]
```diff
@@ -43,6 +43,8 @@ limitations under the License.
 
 namespace xla {
 
+inline constexpr int64_t kPersistentTempMemorySpaceColor = 100;
```
Comment (tdanyluk) [inline xla/service/buffer_assignment.h:46]: Should this be next to kInvalidColor?
Why was the value 100 chosen? Are we sure that it doesn't collide with other values?

**Review by tdanyluk — COMMENTED** [review]
Thanks, could you please elaborate what is the use case for this?

File: xla/service/buffer_assignment.h (hunk 6 above)
Comment (shawnwang18) [reply xla/service/buffer_assignment.h:46]: No particular reason for chosing 100 as the number, just randomly select a value that is not collide with others. Yes, it should be better that XLA put all available ids in a centralized places, but nccl memory space id was hardcoded in the code. Can we make this as TODO that will be fixed in a separate PR?

File: xla/xla.proto (hunk 5 above)
Comment (shawnwang18) [reply xla/xla.proto]: I changed the flag name to xla_gpu_temp_buffer_use_cuda_async_allocator, it requires to use the `cudaAsyncAllocator` because it supports using separate memory pool for allocation, which will just ensure that the allocated address happens to be the stable.

File: xla/service/buffer_assignment.h (hunk 4 above)
Comment (shawnwang18) [reply xla/service/buffer_assignment.h]: Currently this feature is only valid for CUDA devices. but in theory it should also work for ROCM or other devices, as long as we use a separate memory allocator for temp buffer. BufferAssignment pass is a general optimization pass, so currently I made a patch here.

File: xla/service/buffer_assignment.cc (hunk 3 above)
Comment (shawnwang18) [reply xla/service/buffer_assignment.cc]: For temp buffer with new memory space color, it will no longer be valid. I changed the assertion like below: 
  
  if (!(IsPreallocatedTempBuffer() &&
        color() == kPersistentTempMemorySpaceColor)) {
    CHECK_EQ(buffer.color(), color())
        << "Buffer color " << buffer.color() << " for buffer " << buffer
        << " does not match allocation color " << color() << ".";
  }

File: xla/pjrt/gpu/se_gpu_pjrt_client.cc (hunk 2 above)
Comment (shawnwang18) [reply xla/pjrt/gpu/se_gpu_pjrt_client.cc:801]: Thanks, async allocator was merged in a separate PR.

File: xla/debug_options_flags.cc (hunk 1 above)
Comment (shawnwang18) [reply xla/debug_options_flags.cc:1269]: fixed

File: xla/service/buffer_assignment.cc [hunk 7]
```diff
@@ -2108,7 +2124,12 @@ BufferAssigner::CreateAssignment(
   for (const auto& [color, computations] : private_stacks) {
     private_stack_colors.insert(color);
   }
-  assignment->CombineTempAllocations(private_stack_colors);
+  const bool temp_buffer_use_async_allocator =
```
Comment (ezhulenev) [inline xla/service/buffer_assignment.cc:2127]: This should not be here as well, this should be passed from GPU compiler via arguments

File: xla/service/buffer_assignment.h [hunk 8]
```diff
@@ -43,6 +43,8 @@ limitations under the License.
 
 namespace xla {
 
+inline constexpr int64_t kTempBufferUsingCudaAsyncAllocatorColor = 100;
```
[the remark this answers is not in the document]
Comment (shawnwang18) [reply xla/service/buffer_assignment.h]: Fixed

File: xla/service/buffer_assignment.cc (hunk 7 above)
Comment (shawnwang18) [reply xla/service/buffer_assignment.cc]: Fixed

**Comment by tdanyluk** [comment]
I'm withdrawing from this CL, deferring to ezhulenev and cheshire

**Review by ezhulenev — APPROVED** [review]

File: xla/service/gpu/gpu_memory_space_assignment.h [hunk 9]
```diff
@@ -29,6 +29,7 @@ namespace xla {
 namespace gpu {
 
 inline constexpr int64_t kCollectiveMemorySpaceColor = 1;
+inline constexpr int64_t kTempBufferMemorySpaceColor = 5;
```
[the remark this answers is not in the document]
Comment (shawnwang18) [reply xla/service/gpu/gpu_memory_space_assignment.h]: 0,1,5 should have be taken. I fixed it with value 2. 

Maybe best to another PR that puts all the memory space color value in a single place.

File: xla/service/buffer_assignment.cc [hunk 10]
```diff
@@ -700,6 +703,17 @@ void BufferAssignment::CombineTempAllocations(
             temp_allocation.peak_buffers_.begin(),
             temp_allocation.peak_buffers_.end());
       }
+
+      if (temp_buffer_color.has_value()) {
+        // If the temp allocation does not have special color requirement, and
+        // we assign it to a new color type indicating that it will be allocated
+        // with separate allocator to persistent allocation.
+        VLOG(3) << "Using async allocator for temp allocations, size "
```
[the remark this answers is not in the document]
Comment (shawnwang18) [reply xla/service/buffer_assignment.cc]: fixed

**Review by ezhulenev — APPROVED** [review]

**Review by penpornk — CHANGES_REQUESTED** [review]
We have a new policy requiring test updates when adding new features or making any behavioral changes. Could you please help add a test for this?

**Review by ezhulenev — APPROVED** [review]

**Review by ezhulenev — APPROVED** [review]

**Comment by reedwm** [comment]
This is failing the test `pjrt_c_api_gpu_test.cc` internally since GOOGLE_CUDA and CUDA_VERSION are not defined in `se_gpu_pjrt_client.cc`, causing the error `CUDA async allocator requires CUDA >= 11.2`.

I'm not sure why GOOGLE_CUDA and CUDA_VERSION are not defined, nor how to fix this. I'm asking internally to see if anyone else knows.
